### ISSUE TITLE
Segfault on startup on Ubuntu 22.10

### ISSUE THREAD

**Issue by d-torrance** [Macaulay2/M2#2604]
I've noticed that the amd64 PPA build for Ubuntu 22.10 (the current development release -- not to be officially released until October) has been segfaulting for the last week or so (e.g., [this build](https://launchpadlibrarian.net/622079032/buildlog_ubuntu-kinetic-amd64.macaulay2_1.20.0.1+git202209061534-0ppa202209021435~ubuntu22.10.1_BUILDING.txt.gz)).

I was able to reproduce the issue locally after building in an Ubuntu 22.10 chroot.  Here's the stacktrace:

```m2
(kinetic-amd64)root@mixtuppa:~/macaulay2# M2/usr-dist/x86_64-Linux-Ubuntu-22.10/bin/M2 --no-core --notify
-- loading startup.m2 for the first time
-- phase 1
-- executable = /root/macaulay2/M2/usr-dist/x86_64-Linux-Ubuntu-22.10/bin/M2-binary
Macaulay2, version 1.20.0.1
-- prefixDirectory = /root/macaulay2/M2/usr-dist/
-- topBuilddir     = /root/macaulay2/M2/
-- SIGSEGV
-* stack trace, pid: 2306763
 0# stack_trace(std::ostream&, bool) at ../d/main.cpp:127
 1# segv_handler at ../d/main.cpp:244
 2# 0x00007F2136671D30 in /lib/x86_64-linux-gnu/libc.so.6
 3# evaluate_evalraw at ../d/evaluate.d:1469
 4# evaluate_applyFCC at ../d/evaluate.d:563
 5# evaluate_evalraw at ../d/evaluate.d:1305
 6# evaluate_applyFCC at ../d/evaluate.d:563
 7# evaluate_evalraw at ../d/evaluate.d:1305
 8# evaluate_evalraw at ../d/evaluate.d:1269
 9# evaluate_evalraw at ../d/evaluate.d:1269
10# evaluate_evalraw at ../d/evaluate.d:1369
11# evaluate_evalraw at ../d/evaluate.d:1344
12# evaluate_evalraw at ../d/evaluate.d:1269
13# evaluate_applyFCCS at ../d/evaluate.d:845
14# evaluate_evalraw at ../d/evaluate.d:1305
15# evaluate_evalexcept at ../d/evaluate.d:1474
16# readeval3(parse_TokenFile_struct*, char, parse_DictionaryClosure_struct*, char, char, char) at ../d/interp.dd:271
17# readeval(parse_TokenFile_struct*, char, char) [clone .lto_priv.0] at ../d/interp.dd:284
18# interpFunc(ArgCell*) at ../d/main.cpp:194
19# SupervisorThread::threadEntryPoint(void*) at ../system/supervisor.hpp:100
20# GC_inner_start_routine in /lib/x86_64-linux-gnu/libgc.so.1
21# GC_call_with_stack_base in /lib/x86_64-linux-gnu/libgc.so.1
22# 0x00007F21366C62E2 in /lib/x86_64-linux-gnu/libc.so.6
23# 0x00007F2136755240 in /lib/x86_64-linux-gnu/libc.so.6
-- end stack trace *-
```

Usually, these command line options would give us the following:

```m2
$ M2 --no-core --notify
-- loading startup.m2 for the first time
-- phase 1
-- executable = /usr/bin/M2-binary
Macaulay2, version 1.20.0.1
-- prefixDirectory = /usr/
-- topBuilddir     = 
-- topSrcdir       = 
-- phase 2
   -- phase 3
   -- phase 4
```

Comparing the two, it looks like the segfault must be happening around here:

https://github.com/Macaulay2/M2/blob/ccb257c2ad7e909ecc174096079798f5fe062a41/M2/Macaulay2/m2/startup.m2.in#L353-L357

**Comment by DanGrayson** [comment]
But the stack trace says the segmentation fault is happening at "evaluate_evalraw at ../d/evaluate.d:1469".  Can you reproduce the bug while compiling in debug mode?  Then you could get more info using gdb.

**Comment by d-torrance** [comment]
I've confirmed that the segfault happens during the call to `get` on line 355.  I'll try compiling in debug next.

**Comment by DanGrayson** [comment]
What does the file "srcdir" contain?

**Comment by d-torrance** [comment]
No segfault in debug mode.  :(

> What does the file "srcdir" contain?

```
(kinetic-amd64)root@mixtuppa:~/macaulay2/M2# cat ./srcdir 
./
```

**Comment by d-torrance** [comment]
Two things that are probably related:

* Ubuntu 22.10 defaults to GCC 12 (the earlier releases default to GCC <= 11)
* Ubuntu defaults to builds with link-time optimization (the GCC <= 11 Ubuntu builds with LTO work file, and the Debian GCC 12 builds *without* LTO also work fine)

**Comment by DanGrayson** [comment]
What is link time optimization?

**Comment by d-torrance** [comment]
Here's Ubuntu's wiki page on the subject: https://wiki.ubuntu.com/ToolChain/LTO

Fedora is using it, too, and Debian plans to switch to it in the not-too-distant future.

**Comment by DanGrayson** [comment]
Well, that doesn't sound bad.

If you give up, let me know how to install a copy of that ubuntu version and I'll take a look.

**Comment by d-torrance** [comment]
> If you give up, let me know how to install a copy of that ubuntu version and I'll take a look.

Yeah, I'm pretty stumped.  The segfault is happening somewhere between the top-level call to `get` in `startup.m2.in` and the interpreter-level `getfun` in `actors4.d`, so I guess it makes sense that we're getting `evaluate.d` stuff in the stack trace.

My guess is that you wouldn't necessary need Ubuntu 22.10 to reproduce the issue, but just try compiling  using GCC 12 with the LTO build flags.  From the wiki link above:

> Enabling lto adds the flags `-flto=auto -ffat-lto-objects` to `CFLAGS`, `CXXFLAGS`, `OBJCFLAGS` ... and the flag `-flto=auto` to LDFLAGS.

But a quick way to get an Ubuntu 22.10 chroot on another Ubuntu system is to install `ubuntu-dev-tools` and run `mk-sbuild kinetic`.

**Comment by DanGrayson** [comment]
Are you saying that 22.04 might display the bug?

**Comment by d-torrance** [comment]
> Are you saying that 22.04 might display the bug?

Yes -- just confirmed this by building the development branch in 22.04 using GCC 12 and the LTO build flags and got the segfault.  Here was my call to the `configure` script:

```
../../configure --with-system-gc --with-system-memtailor --with-system-mathic --with-system-mathicgb --with-python CC=gcc-12 CXX=g++-12 CFLAGS="$(dpkg-buildflags --get CFLAGS)" CXXFLAGS="$(dpkg-buildflags --get CXXFLAGS)" LDFLAGS="$(dpkg-buildflags --get LDFLAGS)"
```
Note that `dpkg-buildflags` is in the `dpkg-dev` package.  I imagine you could skip the various `--with-*` flags -- I just use those so I don't have to build all the submodules.

**Comment by d-torrance** [comment]
I've confirmed that disabling the LTO build flags fixes the PPA build for Ubuntu 22.10.  Today's build has gotten past all that and is happily building documentation.

**Comment by DanGrayson** [comment]
Oh, well, then, maybe I don't have to do anything, since it's clearly a bug in the LTO, not in our code.

**Comment by mahrud** [comment]
Fedora uses GCC 12 and enables LTO by default, I don't have any issues.

**Comment by DanGrayson** [comment]
Mahrud and Doug, maybe if we compare the exact gcc versions on your two machines, we can see whether the LTO bug has been fixed.

**Comment by d-torrance** [comment]
On my Ubuntu 22.04 machine:
```
$ gcc-12 --version
gcc-12 (Ubuntu 12-20220319-1ubuntu1) 12.0.1 20220319 (experimental) [master r12-7719-g8ca61ad148f]
```

And the 22.10 chroot:
```
$ gcc --version
gcc (Ubuntu 12.2.0-1ubuntu1) 12.2.0
```
@mahrud - Are you using the same LTO flags? `-flto=auto -ffat-lto-objects` in `CFLAGS`/`CXXFLAGS` and `-flto=auto` in `LDFLAGS`?

**Comment by jkyang92** [comment]
I've managed to reproduce this, and I think this is a linker bug. In particular, in the lto version, `stdio_get` seems to be inlined incorrectly into `getfun`.

In particular in the inlined version, there is a `type_` field in the return value of `stdio_get` that is never set, causing a switch in `getfun` to not take any case, causing a null value. Here's my guess on what happened.

1. The LTO inlines `stdio_get` into `getfun`
2. It realizes that the switch after stdio_get in `getfun` can be optimized away, and that the `type_` field in the return value of `stdio_get` is then never used
3. But it then jumps to before the switch rather than into the correct case.

As evidence, this is basically what happened to the case where stdio_get returns an error, except in that case, the linker jumps to the correct branch of the case directly.

For reference, I'm on Arch with gcc 12.2.1.

**Comment by d-torrance** [comment]
> **Written by Claude** (Claude Opus 5, via Claude Code), not by @d-torrance, whose account posted it -- please weigh it accordingly.

Still broken with a newer compiler, and this time with a same-tree control.

The thread only ever tested GCC 12 — 12.0.1 and 12.2.0 here, 12.2.1 for @jkyang92 — so Dan's
question about comparing versions to see whether the LTO bug had been fixed was never answered.
It has not been. On GCC 13.3.0 / Ubuntu 24.04, building today's development tree twice from the
same source with `--with-system-libs` and identical `dpkg-buildflags`, differing only in whether
`-flto=auto -ffat-lto-objects` was stripped:

```
LTO:      -- topBuilddir     = .../lto-build/
          -- SIGSEGV
          make: *** [.../Core/tvalues.m2] Error 1

control:  -- topBuilddir     = .../nolto-build/
          -- topSrcdir       = .../M2/M2/
          -- coredir         = .../share/Macaulay2/Core/
          Macaulay2, version 1.26.06
          -- phase 2 / 3 / 4        exit 0
```

It dies exactly where you pinned it in 2022, at the `get(topBuilddir|"srcdir")` call — now
`startup.m2.in:347-348`. The stack trace matches the original frame for frame:
`evaluate_evalraw` / `applyFCC` / `evalraw` / `applyFCC` / `evalraw` … `applyFCCS` / `evalexcept` /
`readeval3` / `readeval`, with `readeval` still carrying the `[clone .lto_priv.0]` marker. Only the
line numbers moved with the source. Nothing has changed on either side of the interaction: `getfun`
is still `actors4.d:728`, the `get` call is still where it was, and there is no LTO handling anywhere
in `configure.ac` or the cmake files.

It has also recurred independently. #3765 (Arch, May 2025) is the same crash at the same
`tvalues.m2` step with the same `.lto_priv.0` marker, and was resolved the same way — by turning LTO
off in the PKGBUILD, not by fixing anything.

What this does not explain is why @mahrud's Fedora builds were fine with GCC 12 and LTO on; that
question is still open, and the answer probably matters for narrowing this down to something
reportable upstream.

Worth noting that the workaround is load-bearing rather than historical: `dpkg-buildflags` still
emits `-flto=auto -ffat-lto-objects` by default here, so anything that builds M2 with distribution
defaults and does not override them gets the segfault.

<sub>Triaged in a sweep of the open issues; source of truth: [`bug-triage/issues.tsv`](https://github.com/d-torrance/M2/blob/bug-triage/bug-triage/issues.tsv).</sub>
